### PR TITLE
Task Manager: Implement window grouping option

### PR DESCRIPTION
**Pull request by jleclanche**
This is a cleaned up version of #109.

Things that need to happen:

* Remove the eye candy stuff. It's nasty. It isn't written well and it doesn't respect any of the compositing effect options or anything like that so most of the time it's out of place.
* This stuff needs to be fixed: 
image
* "Close on middle click" in the task manager settings needs to be moved to "Taskbar Contents"

Anyway, seeing the state of this, and seeing that the original author is no longer available to work on this, I think we have to drop it from 0.9. 

Fixes lxde/lxqt#132.

### CODE REVIEW

**Comment by jleclanche** [comment]
Additionally, no grouping window should show for single items.

**Comment by paulolieuthier** [comment]
Also, the configuration options should be handled similarly to how it's done in the other plugins.

**Comment by luis-pereira** [comment]
@jleclanche Helix deletes not only the branch containing the commit but also others :(

**Comment by jleclanche** [comment]
@luis-pereira yeah i noticed this before. Ill try to look into it when I get back; I'm not sure why it happens right now.

**Comment by PCMan** [comment]
Why Helix bot automatically delete this branch?

**Comment by luis-pereira** [comment]
@PCMan It's some defect. It happens to me too. @jleclanche is aware and will fix it.

**Comment by palinek** [comment]
@jleclanche this one still needed to finish? is anybody working on in?

**Comment by paulolieuthier** [comment]
@palinek feel free to work on it.

**Comment by palinek** [comment]
for me (running on Frost theme) I needed to add ```color``` style for active taskbar button e.g.:
```css
#TaskBar QToolButton:on{
    color: #DEDEDE;
    background: #1A1A1A;
    border-left: 3px solid #1A80B6;
}
```

**Comment by palinek** [comment]
@paulolieuthier
The commit 4c63f1b seems to solve my problem with ad-hoc SEGFAULTS.
On IRC you mentioned next needed changes for the groupping taskbar behavior & configuration. I can't remeber all of them. Please, put it here.

**Comment by paulolieuthier** [comment]
@palinek I just commited most of it, but I broke the click behaviour. The popup doesn't appear anymore. Please, fix it. I'll be back today.

**Comment by palinek** [comment]
Actualy there are some problems with the taskbar.
- on vertical panel (and taskbar not autorotated) there is always free space on top of the buttons:
screenshot

- on vertical panel (and taskbar autorotated) the popup is wronq (i thing it should be either autorotated or not taking height from height of group button):
screenshot

**Comment by paulolieuthier** [comment]
Hey! It's working very well now. It's just very, very ugly :smile:. I'll clean up the commit history and wait until more people test it.

**Comment by paulolieuthier** [comment]
@palinek if you think it still needs clean up, redesign or simple refactoring, that's the time.

**Comment by paulolieuthier** [comment]
Actually, I'm doing a large refactoring and optimizations, will commit tomorrow.

**Comment by paulolieuthier** [comment]
Dragging now broke. That also needs refactoring. Will do it today or tomorrow.

**Comment by paulolieuthier** [comment]
Drag and drop for taskgorups on taskbar and taskbuttons on popup now restored. There are still a few bugs, there. Plus the bugs @palinek reported in the previous comments.

**Comment by paulolieuthier** [comment]
@palinek please help me test it. All drag and drop should be working now. I mean DND inside the popups, DND of a file to a window (dragging to a button must make the button's window open) and of course the normal DND of taskgroups in the taskbar.

**Comment by paulolieuthier** [comment]
@jleclanche @palinek @luis-pereira @kubanecxxx I think it's ready now. Please test it and report bugs/crashes. If it's ok, I'll rebase, squash and merge it.

**Comment by paulolieuthier** [comment]
@jleclanche @palinek @luis-pereira @kubanecxxx I think it's ready now. Please help me testing and finding bugs. Then I'll rebase, squash and merge it.

**Comment by jleclanche** [comment]
Ill test it this week I hope.

**Comment by palinek** [comment]
I'll try to test this night..

**Comment by palinek** [comment]
Just tested
- moving buttons in taskbar works OK (just a minor "bug" is that after droping the button one or more buttons stay highlighted (:hover))
- drag&drop of file into window in a groupped button -> the popup stays open (after application window activated) and mouse leaves the popup
- just moving the mouse over groupped buttons -> multiple popups are shown for a short period of time
- on vertical panel with autorotation not enabled -> "one empty row on top" bug still remainig

**Comment by palinek** [comment]
@paulolieuthier I have removed the delay for timer and this helped for multiple popup showing. Now I don't see the reason for having a timer for closing the popup... We wanted to remove all eyecandy stuff, not?

**Comment by paulolieuthier** [comment]
@palinek the timer for closing the popup is not a eyecandy feature, it's a usability thing. You don't want the popup to closing all the time if you accidentally leave it by 1px. Do you see? No problem in reducing the time, but it shouldn't be zero.

**Comment by palinek** [comment]
@paulolieuthier well then we need to add logic for immediate hide of "previous" popup when the "new" popup is risen

**Comment by palinek** [comment]
...now only reamining bug is the "style" issue while moving buttons around the taskbar (but this is known one and also present in current master)

File: plugin-taskbar/lxqttaskbutton.cpp [hunk 1]
```diff
@@ -81,6 +90,10 @@ LxQtTaskButton::LxQtTaskButton(const WId window, QWidget *parent) :
 
     updateText();
     updateIcon();
+
+    mDNDTimer->setSingleShot(true);
+    mDNDTimer->setInterval(1000);
```
Comment (palinek) [inline plugin-taskbar/lxqttaskbutton.cpp]: Isn't this too much? IMHO half a second is enough...

Comment (paulolieuthier) [inline plugin-taskbar/lxqttaskbutton.cpp]: 750ms and we're even :smile:

**Comment by paulolieuthier** [comment]
Just fixed those (from comments).

**Comment by paulolieuthier** [comment]
Please test.

**Comment by paulolieuthier** [comment]
Anyone?

**Comment by luis-pereira** [comment]
@paulolieuthier can you rebase ?

**Comment by paulolieuthier** [comment]
@luis-pereira rebased.

**Comment by paulolieuthier** [comment]
Merging to get more feedback.

**Comment by palinek** [comment]
@paulolieuthier upatient ? me too :-D

**Comment by paulolieuthier** [comment]
@palinek yes haha. The master branch is where new features (not so much tested, but working) land.

**Comment by paulolieuthier** [comment]
@intialonso now it's your time. Grouping is very ugly.

**Comment by luis-pereira** [comment]
Congratulations to all the people that work on this. Special thanks to @kubanecxxx

**Pull request merged**